### PR TITLE
Mlgym#193 sys info endpoint

### PR DESCRIPTION
**Pull request by moinam**
Added run_config to event storage using existing API functionality.
New API to fetch system info (CPU & GPU):
_/system-info_
Response:
{
    "platform": <Operating system>,
    "platform-release": <OS version>,
    "architecture": <system architecture ex: AMD64>,
    "processor": <processor name>,
    "ram": <ram in GB>,
    "CUDNN_version":<cudnn version>,
    "num_cuda_device": <number of cuda devices>,
    "cuda_device_list": [
        {
            "name": <name of GPU>,
            "multi_proc_count": <Processor count of GPU>,
            "total_memory": <Total GPU memory in GB>
        }
    ]
}

### CODE REVIEW

File: src/ml_board/README.md [hunk 1]
```diff
@@ -70,6 +70,31 @@ payload:
 }
 ```
 
+**Model Card**
+
+Fetch System details like CPU & GPU for model card.
+
+_GET /system-info_
```
Comment (le1nux) [inline src/ml_board/README.md:77]: I think we should have this information for each experiment in a  gridsearch. We could add it to the logging files like `<gridsearch_id>/<experiment_id>/environment.json`. 

You are querying this information on the fly for each request. Note that the API server might be running not on the same node as the MLgym computation part.

Comment (le1nux) [reply src/ml_board/README.md:77]: Additional information would be helplful as well such as python version and the installed python packages.

File: src/ml_board/backend/restful_api/data_access.py [hunk 2]
```diff
@@ -406,3 +413,34 @@ def delete_checkpoint_resource(
...
         else:
             raise FileNotFoundError(f"File in path {requested_full_path} not found.")
+
+    def get_system_info(self):
+        """
+        ``HTTP GET`` Fetch System Information for model card.
+
+        :returns: JSON object - System Information of host machine (CPU & GPU)
+        """
+        try:
```
Comment (le1nux) [inline src/ml_board/backend/restful_api/data_access.py:423]: this information should not be retrieved upon request. Let's store it the way, as explained above.

File: src/ml_gym/cmd_entrypoint/cmd.py [hunk 3]
```diff
@@ -56,81 +64,110 @@ class LoggingConfiguration:
...
+def get_gym_from_environment_config(
+    env_config: Union[MultiProcessingEnvironmentConfig, MainProcessEnvironmentConfig, AccelerateEnvironmentConfig],
+    logger_collection_constructable: LoggerConstructableIF,
+    gs_restful_api_client_constructable: GridSearchAPIClientConstructableIF,
+    num_epochs: int,
+    num_batches_per_epoch: int = None,
+    accelerator: Accelerator = None,
+) -> Gym:
     if isinstance(env_config, MainProcessEnvironmentConfig):
-        gym = GymFactory.get_sequential_gym(logger_collection_constructable=logger_collection_constructable,
```
Comment (le1nux) [inline src/ml_gym/cmd_entrypoint/cmd.py:68]: I see all these code reformatting changes. Could you please disable those in your IDE and undo these changes please?

**Review by le1nux — CHANGES_REQUESTED** [review]
This PR needs some changes, as described in the comments. 

https://github.com/mlgym/mlgym/blob/dev/src/ml_gym/gym/gym_jobs/gym_job.py#L31
you can log the environment via the `gs_api_client` for each experiment in the `GymJob` instantiation and then store the information as a JSON in the MLgym backend (like the config for the individual experiments).

File: src/ml_board/README.md (hunk 1 above)
Comment (moinam) [reply src/ml_board/README.md:77]: > I think we should have this information for each experiment in a gridsearch. We could add it to the logging files like `<gridsearch_id>/<experiment_id>/environment.json`.
> 
> You are querying this information on the fly for each request. Note that the API server might be running not on the same node as the MLgym computation part.

this is done

File: src/ml_board/backend/restful_api/data_access.py (hunk 2 above)
Comment (moinam) [reply src/ml_board/backend/restful_api/data_access.py]: done

File: src/ml_gym/cmd_entrypoint/cmd.py (hunk 3 above)
Comment (moinam) [reply src/ml_gym/cmd_entrypoint/cmd.py]: done

File: src/ml_gym/persistency/io.py [hunk 4]
```diff
@@ -277,6 +283,38 @@ def get_experiment_statuses(self, grid_search_id: str) -> List[ExperimentStatus]
         response = GridSearchRestfulAPIClient._get_json_resource(url)
         experiment_statuses = [ExperimentStatus(**r) for r in response]
         return experiment_statuses
+    
+    def create_system_info(self) -> Dict:
```
Comment (le1nux) [inline src/ml_gym/persistency/io.py:287]: I think we should move this to a class like `Environment`, with a static function e.g. in the util package
https://github.com/mlgym/mlgym/tree/master/src/ml_gym/util

```python 
class Environment: 
    @static_method
    create_system_info() -> Dict:
        ....

```

Comment (le1nux) [reply src/ml_gym/persistency/io.py:287]: I think we should include all the package names (even if it's a long list, we can just filter it in the frontend) and the original python CMD call. Maybe we could check the entire process command?

**Review by le1nux — CHANGES_REQUESTED** [review]
Minor change suggestions. Otherwise, looking good :)

File: src/ml_gym/util/util.py [hunk 5]
```diff
@@ -16,8 +16,48 @@
...
+                    dev_list.append(
+                        {
+                            "name": torch.cuda.get_device_name(i),
+                            "multi_proc_count": torch.cuda.get_device_properties(i).multi_processor_count,
+                            "total_memory": f"{round(torch.cuda.get_device_properties(i).total_memory / 1e9, 2)} GB",
+                        }
+                    )
+                info["cuda_device_list"] = dev_list
+            return info
+        except:
```
Comment (le1nux) [inline src/ml_gym/util/util.py:59]: better: 

```python

except Exception as e:
    raise SystemInfoFetchError(f"Unable to fetch System Info") from e


```

**Review by le1nux — APPROVED** [review]
minor suggestion. otherwise looks good :) can be merged
